File: cpp/dp/digit_dp.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
struct DigitDP
{
  string s;
  int mod;
  vector<vector<array<ll, 2>>> memo;
  ll dfs(int pos, int rem, bool tight, bool started)
  {
    if (pos == (int)s.size())
      return started && rem == 0;
    if (!tight && memo[pos][rem][started] != -1)
      return memo[pos][rem][started];
    int lim = tight ? s[pos] - '0' : 9;
    ll ans = 0;
    for (int d = 0; d <= lim; d++)
    {
      bool ns = started || d;
      ans += dfs(pos + 1, ns ? (rem + d) % mod : 0, tight && d == lim, ns);
    }
    if (!tight)
      memo[pos][rem][started] = ans;
    return ans;
  }
  ll solve(long long N, int m)
  {
    if (N < 0)
      return 0;
    s = to_string(N);
    mod = m;
    memo.assign(s.size(), vector<array<ll, 2>>(mod, {-1, -1}));
    return dfs(0, 0, 1, 0);
  }
};

File: python/dp/digit_dp.py
from functools import lru_cache


def count_upto(n):
    digits = list(map(int, str(n)))

    @lru_cache(None)
    def dp(pos, tight, started, rem):
        if pos == len(digits):
            return int(started and rem == 0)
        limit = digits[pos] if tight else 9
        ans = 0
        for d in range(limit + 1):
            nt = tight and d == limit
            ns = started or d != 0
            ans += dp(pos + 1, nt, ns, (rem + d) % 3 if ns else 0)
        return ans

    return dp(0, True, False, 0)
